doc/sphinx/proofs/writing-proofs/rewriting.rst

```diff
@@ -883,14 +781,129 @@ the conversion in hypotheses :n:`{+ @ident}`.
+described first.
+
+.. cmd:: Opaque {+ @reference }
+
+   This command accepts the :attr:`global` attribute.  By default, the scope
+   of :cmd:`Opaque` is limited to the current section or module.
+
+   This command has an effect on unfoldable constants, i.e. on constants
+   defined by :cmd:`Definition` or :cmd:`Let` (with an explicit body), or by a command
+   associated with a definition such as :cmd:`Fixpoint`, etc,
+   or by a proof ended by :cmd:`Defined`. The command tells not to unfold the
+   constants in the :n:`@reference` sequence in tactics using δ-conversion (unfolding
+   a constant is replacing it by its definition).
+
+   :cmd:`Opaque` has also an effect on the conversion algorithm of Coq, telling
+   it to delay the unfolding of a constant as much as possible when Coq
+   has to check the conversion (see Section :ref:`conversion-rules`) of two distinct
+   applied constants.
+
+.. cmd:: Transparent {+ @reference }
+
+   This command accepts the :attr:`global` attribute.  By default, the scope
+   of :cmd:`Transparent` is limited to the current section or module.
+
+   This command is the converse of :cmd:`Opaque` and it applies on unfoldable
+   constants to restore their unfoldability after an Opaque command.
+
+   Note in particular that constants defined by a proof ended by Qed are
+   not unfoldable and Transparent has no effect on them. This is to keep
+   with the usual mathematical practice of *proof irrelevance*: what
+   matters in a mathematical development is the sequence of lemma
+   statements, not their actual proofs. This distinguishes lemmas from
+   the usual defined constants, whose actual values are of course
+   relevant in general.
+
+   .. exn:: The reference @qualid was not found in the current environment.
+
+      There is no constant named :n:`@qualid` in the environment.
+
+.. seealso:: :ref:`performingcomputations` and :ref:`proof-editing-mode`
+
+.. _vernac-strategy:
+
+.. cmd:: Strategy {+ @strategy_level [ {+ @reference } ] }
+
+   .. insertprodn strategy_level strategy_level_or_var
+
+   .. prodn::
+      strategy_level ::= opaque
+      | @integer
+      | expand
+      | transparent
+      strategy_level_or_var ::= @strategy_level
+      | @ident
+
+   This command accepts the :attr:`local` attribute, which limits its effect
+   to the current section or module, in which case the section and module
+   behavior is the same as :cmd:`Opaque` and :cmd:`Transparent` (without :attr:`global`).
+
+   This command generalizes the behavior of the :cmd:`Opaque` and :cmd:`Transparent`
+   commands. It is used to fine-tune the strategy for unfolding
+   constants, both at the tactic level and at the kernel level. This
+   command associates a :n:`@strategy_level` with the qualified names in the :n:`@reference`
+   sequence. Whenever two
+   expressions with two distinct head constants are compared (for
+   instance, this comparison can be triggered by a type cast), the one
```

Comment:
A more common case than a type cast is that if you have `f : A -> B` and `x : C`, typechecking `f x` will result in converting `A` and `C`.